schematic/base.py

```diff
@@ -0,0 +1,446 @@
+    the built in comparison operators. For two TableColumnTypes, A and B,
+    if A is a superset of B, then A > B evaluates to True.
+
+    Note that TableColumnTypes are only fully comparable when in the
+    same subtree. If A is not found when traversing next_less_restrictive
+    in B and vice versa, A < B, A > B, and A == B will all return False
+
+    Attributes:
+      name: The name of the type
+      next_less_restrictive: The TableColumnType which is next less restrictive
+      parameter: Instance parameter for this class. E.g., '256' for a VARCHAR(256)
+
+    TODO(Cody): Comparisons of incomparable operands should raise an exception
+    TODO(Cody): Factor out the "most to least restrictive" iterator logic into a method on this class.
+    TODO(Cody): Change the "parameter" attribute to use **kwargs.
+    """
+    name = "TableColumnType"
+    next_less_restrictive = None
+    parameterized = False
+
+    def __init__(self, parameter=None):
+        if parameter is not None and not self.parameterized:
+            raise ValueError(
+                "{} is not a parameterized type.".format(
+                    self.name))
+        self.parameter = parameter
+
+    def __hash__(self):
+        return hash((self.name, self.parameter))
+
+    def __repr__(self):
+        if self.parameterized:
+            return "{} ({})".format(self.name, self.parameter)
+        else:
+            return self.name
+
+    def __eq__(self, other):
+        return isinstance(
+            other, type(self)) and other.parameter == self.parameter
+
+    def __lt__(self, other):
+        return other > self
+
+    def __gt__(self, other):
+        nlr = other
+        while nlr:
+            nlr = nlr.next_less_restrictive() if nlr.next_less_restrictive else None
+            if nlr == self:
+                return True
+        return False
+
+    def get_depth(self):
+        """Get the distance between this TableColumnType
+        and the least restrictive TableColumnType in its
+        next_less_restrictive linked list
+
+        Returns:
+          An int
+        """
+        depth = -1
+        nlr = self
+        while nlr:
+            nlr = nlr.next_less_restrictive
+            depth += 1
+        return depth
+
+    def value_is_compatible(self, value):
+        """Checks to see if the given value can be inserted into a column of
+           the type described by this instance.
+        Args:
+          value: The value to check, a string
+        Returns:
+          True if _value_is_compatible_superset(self, value) is True and
+          this type isn't parameterized
+        Raises:
+          NotImplementedError: This should be implemented by subclasses
+                               if subclasses are parameterized.
+        """
+        if self.parameterized:
+            raise NotImplementedError
+        else:
+            return self._value_is_compatible_superset(value)
+
+    def _value_is_compatible_superset(self, value):
```

Comment:
```suggestion
    @abstractmethod
    def _value_is_compatible_superset(self, value):
```
Force subclass to define this